Create eip-3779.md (#3781)
ethereum/EIPs: EIPS/eip-3779.md

---

> eip: 3779

Reviewer: ```suggestion
eip: 3781
```

> title: Safe Control Flow for the EVM

Reviewer: Description field missing.  Filled in a placeholder in the suggestion below, feel free to put whatever you want though.  😄 
```suggestion
title: Safe Control Flow for the EVM
description: Introduces contract validation rules for jumps and stack depth that are applied at deploy time.
```

status: Draft
type: Standards Track
category: Core

> author: Greg Colvin <greg@colvin.org>

Reviewer: All EIPs should have at least one GitHub handle in the author list so we can be sure to have in-platform contact information for at least one person.  Either of the following are acceptable:
```suggestion
author: Greg Colvin (@gcolvin)
```
```suggestion
author: Greg Colvin (@gcolvin), Greg Colvin <greg@colvin.org>
```

discussions-to: https://ethereum-magicians.org/t/eip-3779-safe-control-flow-for-the-evm/6975

> created: 2021-8-30

Reviewer: ```suggestion
created: 2021-08-30
```

> ---

Reviewer: ```suggestion
requires: 3540
---
```

1. Insufficient gas
2. More than 1024 stack items
3. Insufficient stack items
4. Invalid jump destination
5. Invalid instruction

We would like to consider EVM code valid iff no execution of the program can lead to an exceptional halting state, but we must be able to validate code in linear time to avoid denial of service attacks.  So in practice, we can only partially meet these requirements.  Our validation algorithm does not consider the codes data and computations, only its control flow and stack use.  This means we will reject programs with any invalid code paths, even if those paths are not reachable at runtime.   Further, conditions *1* and *2*  Insufficient gas and stack overflow  must in general be checked at runtime.  Conditions *3*, *4*, and *5* cannot occur if the code conforms to the following rules.

#### The Rules

1. `JUMP` and `JUMPI` address only constant, valid `JUMPDEST` instructions.
2. For each instruction in the code the `stack depth` is always the same.
3. The `stack depth` is always positive and at most 1024.

> 

Reviewer: Can these be split into 3 separate EIPs?  Smaller EIPs tend to flow through the process better as people can analyze, think about, and discuss them in isolation.  Also, we have found in the past that EIPs that are broad can often get stuck on someone disagreeing with part of it even when everyone agrees on most of it, and by having separate smaller EIPs the agreeable parts can go through and the disagreeable parts can be delayed for further discussion.

Author: I think these hang together around a single algorithm.  I wouldn't know how to split them up.  Plenty of room for improving this one in ways you suggest -- a this point it's a bit of Frankenstein's monster, cobbled together from pieces of older EIPs.

I mention EIP-615 only because its safety properties are well understood after years of discussion and even formal analysis.  It's important that this EIP makes the same guarantees.

                 return false
             }

             // reset PC to destination of jump 
             PC += jumpdest
             continue
         }
         if instruction == JUMPI {

             // check for valid destination
             jumpdest = stack[SP]
             if bytecode[jumpdest] != JUMPDEST {
                 return false
             }

             // reset PC to destination of jump 
             PC += jumpdest

             // recurse to jump to code to validate 
             if !validate(stack[SP])) {
                 return false
             }
             continue 
         }
         
        // effect of instruction on stack
         SP -= remove_items(instruction)
         SP += add_items(instruction)
         if SP < 0 || 1024 < SP {
             return false
         }

         // successful validation of path
         if     PC > code_len ||
                instruction == STOP || 
                instruction == RETURN ||
                instruction == SUICIDE {
             return true
         }

         // advance PC according to instruction
         PC = advance_pc(PC, instruction)
      }    
   }
```

> ## Rationale

Reviewer: Consider moving some of the content from the specification into the rationale.  Anything that explains *why* a particular thing is made should be in the rationale, while the specification should be strictly the things required to implement this change.

I like to think of the specification like a recipe book, where the author doesn't explain *why* you are adding cumin to the dish, they just tell you exactly how much cumin to add.

The alternative to checking validity at creation time is checking it at runtime.  This hurts performance and is a DoS surface.

## Backwards Compatibility

> These changes affect the semantics of existing EVM code – the [EVM Object Format deprecated dynamic use of JUMP and JUMPI.  Contracts that make only static use of JUMP and JUMPI remain valid.  Most contracts to not use dynamic jumps.

Reviewer: I suspect this is what you meant.
```suggestion
These changes affect the semantics of existing EVM code – the [EVM Object Format](./eip-3540.md) deprecated dynamic use of JUMP and JUMPI.  Contracts that make only static use of JUMP and JUMPI remain valid.  Most contracts to not use dynamic jumps.
```

## Security Considerations

> Safety and amenability to static analysis of valid programs are comparable to [EIP-615](https://eips.ethereum.org/EIPS/eip-615),  without imposing syntactic constraints.  Validity can ensured by following the rules given previously, and programs can be validated with the provided algorithm.  The validation algorithm is simple and bounded by the size of the code, allowing for validation at creation time and preventing DoS attacks.  Compilers can easily follow the rules, and in most cases already do.

Reviewer: Copyright section is required, and EIP links need to be relative.

Note: As mentioned previously, since this EIP references 615 its progress through the process will be held up by it.  If you think 615 will make it through first then that is fine, but if you think 615 may stall out or be withdrawn I recommend extracting the necessary information from it and making this EIP standalone.

```suggestion
Safety and amenability to static analysis of valid programs are comparable to [EIP-615](./eip-615.md),  without imposing syntactic constraints.  Validity can ensured by following the rules given previously, and programs can be validated with the provided algorithm.  The validation algorithm is simple and bounded by the size of the code, allowing for validation at creation time and preventing DoS attacks.  Compilers can easily follow the rules, and in most cases already do.
 
## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```

            }

            // reset PC to destination of jump 
            PC += jumpdest

            // check for stack underflow and
            // recurse to jump to code to validate
            stack_depth := SP - entry_SP
            if stack_depth < 0 { 
                return false
            }
            --SP
            if !validate(stack[SP]), stack_depth {
                return false
            }

            continue 
        }
 
        // check effect of instruction on stack
        SP -= remove_items(instruction)
        SP += add_items(instruction)
        if SP < 0 || 1024 < SP {
            return false
        }

        // successful validation of path
        if PC > code_len
               || instruction == STOP
               || instruction == RETURN
               || instruction == SUICIDE {
            return true
        }

        // advance PC according to instruction
        PC = advance_pc(PC, instruction)
    }
```

## Rationale

The alternative to checking validity at creation time is checking it at runtime.  This hurts performance and is a DoS vulnerability.  Thus the above rules and accompanying one-pass validation algorithm.

> _Rule 1_ – requiring constant destinations for `JUMP` and `JUMPI` – forbids dynamic jumps.  This rule is already required by `EIP-3540`.  Jump destinations are currently checked at runtime, but static jumps can be validated at creation time.   

Reviewer 2: How is this required by 3540?

Author: It reads "absolite" , and I misread that for "constant".  I will remove  the reference.
> JUMP/JUMPI uses an absolute offset within the code.

https://github.com/ethereum/EIPs/blob/f4f250c211a5d9628ec81c47d15de45107d6acb3/EIPS/eip-3540.md#changes-to-execution-semantics
